sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java

```diff
@@ -235,12 +284,151 @@ public ReadFiles withAvroDataModel(GenericData model) {
+        MessageType requestedSchema = readContext.getRequestedSchema();
+        RecordMaterializer<GenericRecord> recordConverter =
+            readSupport.prepareForRead(conf, fileMetadata, fileSchema, readContext);
+        boolean strictTypeChecking = options.isEnabled(STRICT_TYPE_CHECKING, true);
+        boolean filterRecords = options.useRecordFilter();
+        reader.setRequestedSchema(requestedSchema);
+        MessageColumnIO columnIO =
+            columnIOFactory.getColumnIO(requestedSchema, fileSchema, strictTypeChecking);
+        for (int i = 0; i < reader.getRowGroups().size(); i++) {
+          if (i < tracker.currentRestriction().getFrom()) {
+            reader.skipNextRowGroup();
+            continue;
+          }
+          if (tracker.tryClaim((long) i)) {
+            PageReadStore pages = reader.readNextRowGroup();
+            i += 1;
+            RecordReader<GenericRecord> recordReader =
+                columnIO.getRecordReader(
+                    pages, recordConverter, filterRecords ? filter : FilterCompat.NOOP);
+            GenericRecord read;
+            long current = 0;
+            long totalRows = pages.getRowCount();
+            while (current < totalRows) {
+              read = recordReader.read();
+              outputReceiver.output(read);
+              current += 1;
+            }
+          } else {
+            break;
+          }
+        }
+      }
+
+      private Configuration setConf() throws Exception {
+        Configuration conf = new Configuration();
+        GenericData model = null;
+        if (modelClass != null) {
+          model = (GenericData) modelClass.getMethod("get").invoke(null);
+        }
+        if (model != null
+            && (model.getClass() == GenericData.class || model.getClass() == SpecificData.class)) {
+          conf.setBoolean(AvroReadSupport.AVRO_COMPATIBILITY, true);
+        } else {
+          conf.setBoolean(AvroReadSupport.AVRO_COMPATIBILITY, false);
+        }
+        return conf;
+      }
+
+      @GetInitialRestriction
+      public OffsetRange getInitialRestriction(@Element FileIO.ReadableFile file) throws Exception {
+        if (!file.getMetadata().isReadSeekEfficient()) {
+          ResourceId filename = file.getMetadata().resourceId();
+          throw new RuntimeException(String.format("File has to be seekable: %s", filename));
+        }
+        SeekableByteChannel seekableByteChannel = file.openSeekable();
+        InputFile inputFile = new BeamParquetInputFile(seekableByteChannel);
+        Configuration conf = setConf();
+        ParquetReadOptions options = HadoopReadOptions.builder(conf).build();
+        ParquetFileReader reader = ParquetFileReader.open(inputFile, options);
+        return new OffsetRange(0, reader.getRowGroups().size());
+      }
+
+      @SplitRestriction
+      public void split(@Restriction OffsetRange restriction, OutputReceiver<OffsetRange> out) {
```

Comment:
I will change it into dividing the range into half each time.

Reply:
The `split ` here is for initial split, not for dynamic split. For example, if you have one file(2 row groups in total) as input, after initial split you will get (file, [group1]), (file, [group2]) as the input of your process fn. Usually, we expose an API from IO to let the end user to set desired initial splits.

Reply:
I think Split into half as the initial splits should be fine as our purpose for making it into splittable doFn is for the read process to be faster. User might not need to put in an actual split info as they might not know exact information about the file before they read them.

Reply:
I feel like it may not feasible to hardcode the initial split. @chamikaramj What do you think?

Reply:
Usual pattern for files is to initial split into 64MB blocks and dynamic split further as needed. I believe this is the tried and tested method for existing file-based sources. This might require changing to a tracker that use byte positions (0 to size of file) instead of a tracker that goes from 0 to number of blocks.

Reply:
@boyuanzz do you think we should add an abstract block-based SDF implementation/utility that will be shared between multiple source implementations (for example, Parquet, Avro) similar to [1] for bounded sources ?

[1] https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/io/BlockBasedSource.java

Reply:
May be we should add a BlockBasedTracker that extends OffsetRangeTracker that goes from 0 to size of file and can split at block boundaries but can provide better progress signals (based on element estimates).

Reply:
It seems like what Jiadai is trying to do is taking a file as element and the number of row group as restriction. When SDF.process() is called, it will read all row groups defined in the restriction for one file. If using BlockBasedTracker, the restriction will be the number of blocks. I think the key point here is to define what should be the element and restriction for this source.

Reply:
I have tried to Implement a block tracker which still provide the information of progress in bytes. So the restriction is still block but the progress method built inside the range tracker tracks the estimate progress.

Reply:
Yes, I think we can build a BlockBasedTracker that uses number of blocks for restriction but aware of byte positions. This will allow it to better perform initial splitting and provide a better progress signal based on estimated element positions (by overriding getProgress()).

We do something like this for progress in existing block-based source: https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/io/BlockBasedSource.java#L241